keps/sig-cluster-lifecycle/0029-20180918-kubeadm-phases-beta.md

```diff
@@ -0,0 +1,184 @@
+    /admin          Generates a kubeconfig file for the admin to use and for kubeadm itself
+    /kubelet        Generates a kubeconfig file for the kubelet to use.
+    ...
+  ...
+````
+
+Phases implementation will be designed in order to ensure that phases are reusable across different
+workflows thus allowing e.g. reuse of phase `certs` in both `kubeadm init` and
+`kubeadm join --master` workflows.
+
+Each kubeadm workflow will be defined and managed using an `Executor` that basically is a component
+in charge of of running phases according to the given order; nested phases will be executed
+immediately after their parent phase.
+
+The `Executor` will be instantiated by kubeadm commands, and the executor options exposed as
+kubeadm command flags thus allowing to filter the list of phases to be executed. As a consequence
+the `kubeadm init` UX will be modified from:
+
+```bash
+Usage:
+  kubeadm init [flags]
+```
+
+to
+
+```bash
+Usage:
+  kubeadm init [workflow-flags] [flags]
+```
+
+[workflow-flags] will be a set of optional flags that allows user to select which phase
+of the `kubeadm init` logic to execute (`--only-phases`) or to skip (`--skip-phases`). e.g.
+
+```bash
+kubeadm init --only-phases certs [flags]
+
+kubeadm init --only-phases certs/ca, certs/apiserver [flags]
+
+kubeadm init --skip-phases addons [flags]
+```
+
+By executing `kubeadm init` without [workflow-flags], all the phases which are part
+of the init workflow will be executed, thus ensuring consistency with the
+current `kubeadm init` behavior.
+
+## Graduation Criteria
+
+* To create a periodic E2E test that bootstraps a cluster using phases
+* To document the new user interface for phases in kubernetes.io
+
+## Implementation History
+
+* [#61631](https://github.com/kubernetes/kubernetes/pull/61631) First prototype implementation 
+  (now outdated)
+
+## Drawbacks
+
+By merging phases into kubeadm workflows derives a reduced capability to customize
+the user interface for each phase. More specifically:
+
+- It would not be possible to provide any kind of advice to the user about which
+  flags are relevant for one specific phase (the help will always show all the flags).
+- It would not be possible to add long description and/or examples to each phase
+- It would not be possible to provide additional flags specific for one phase
+  (the flags are shared between init and all the phases).
+- It would not be possible to expose to the users phases which are not part of kubeadm workflows
+  ("extra" phases should be hosted on dedicated commands).
+
+This is considered an acceptable trade-off in light of the benefits of the suggested
+approach.
+
+## Alternatives
+
+It is possible to graduate phases by simply moving corresponding command to first level,
+but this approach provide less opportunities for reducing the effort
+for maintaining phases up to date with the changes of kubeadm.
```

Comment:
are we still planning to move all `kubeadm alpha phase` -> `kubeadm phase` early in 1.13?

Reply:
I hope so, but this is up to the Sig...